sdkutil/event.go

```diff
@@ -87,3 +104,172 @@ func GetString(attrs []sdk.Attribute, key string) (string, error) {
+// field names.
+func ReflectBlockID(id interface{}) *BlockID {
+	b := BlockID{}
+
+	val := reflect.ValueOf(id)
+	typeOfID := val.Type()
+
+	for i := 0; i < val.NumField(); i++ {
+		field := val.Field(i)
+		switch typeOfID.Field(i).Name {
+		case "Owner":
+			if field.Type() == reflect.TypeOf(sdk.AccAddress{}) {
+				x, ok := field.Interface().(sdk.AccAddress)
+				if ok && len(x) == akashHexLen {
+					b.Owner = &x
+				}
+			}
+		case "DSeq":
+			if field.Type() == reflect.TypeOf(uint64(0)) {
+				x, ok := field.Interface().(uint64)
+				if ok {
+					b.DSeq = &x
+				}
+			}
+		case "GSeq":
+			if field.Type() == reflect.TypeOf(uint32(0)) {
+				x, ok := field.Interface().(uint32)
+				if ok {
+					b.GSeq = &x
+				}
+			}
+		case "OSeq":
+			if field.Type() == reflect.TypeOf(uint32(0)) {
+				x, ok := field.Interface().(uint32)
+				if ok {
+					b.OSeq = &x
+				}
+			}
+		case "Provider":
+			if field.Type() == reflect.TypeOf(sdk.AccAddress{}) {
+				x, ok := field.Interface().(sdk.AccAddress)
+				if ok && len(x) == akashHexLen {
+					b.Provider = &x
+				}
+			}
+		default:
+			// Skip this field
+		}
+	}
+	return &b
+}
+
+// FmtBlockID provides a human readable representation of the block chain ID fields.
+func FmtBlockID(owner *sdk.AccAddress, dseq *uint64, gseq *uint32, oseq *uint32, provider *sdk.AccAddress) string {
+	return BlockID{
+		Owner:    owner,
+		DSeq:     dseq,
+		GSeq:     gseq,
+		OSeq:     oseq,
+		Provider: provider,
+	}.String()
+}
+
+// ParseBlockID returns the values from a human readable ID string.
+func ParseBlockID(id string) (*BlockID, error) {
+	b := &BlockID{}
+
+	parts := strings.Split(id, string(filepath.Separator))
+	if len(parts) < 2 {
+		return nil, ErrInvalidParseBlockIDInput
+	}
+
+	if len(parts[blockIDOwnerPosition]) == akashBech32Len {
+		aa, err := sdk.AccAddressFromBech32(parts[blockIDOwnerPosition])
+		if err != nil {
+			return nil, errors.Wrap(ErrParsingBlockID, err.Error())
+		} else if len(aa) > 0 {
+			b.Owner = &aa
+		}
+	} else {
+		return nil, errors.Wrap(ErrInvalidParseBlockIDInput, "owner address bytes invalid")
+	}
+
+	if len(parts[blockIDDSeqPosition]) > 0 {
+		dseq, err := strconv.ParseUint(parts[blockIDDSeqPosition], 10, 64)
+		if err != nil {
+			return nil, errors.Wrap(ErrParsingBlockID, err.Error())
+		}
+		b.DSeq = &dseq
+	} else {
+		return nil, errors.Wrap(ErrInvalidParseBlockIDInput, "empty dseq value")
+	}
+
+	if len(parts) > 2 && len(parts) < 6 {
+		if len(parts[blockIDGSeqPosition]) > 0 {
+			gseq, err := strconv.ParseUint(parts[2], 10, 64)
+			if err != nil {
+				return nil, errors.Wrap(ErrParsingBlockID, err.Error())
+			}
+			x := uint32(gseq)
+			b.GSeq = &x
+		}
+
+		if len(parts[blockIDOSeqPosition]) > 0 {
+			oseq, err := strconv.ParseUint(parts[3], 10, 64)
+			if err != nil {
+				return nil, errors.Wrap(ErrParsingBlockID, err.Error())
+			}
+			x := uint32(oseq)
+			b.OSeq = &x
+		}
```

Comment:
what if `len(parts) == 3`?

All of the `seqs` should never be zero, you probably don't need to check the string length - I think these should all be guarded by `len(parts) > blockFooPosition` rather than `len(parts[blockFooPosition])`.

Reply:
That's a better way to check; done!